Source/WebCore/platform/graphics/Damage.h

```diff
@@ -56,32 +56,84 @@ class Damage {
         return invalidDamage;
     }
 
-    ALWAYS_INLINE const Region& region() const { return m_region; }
-    ALWAYS_INLINE IntRect bounds() const { return m_region.bounds(); }
-    ALWAYS_INLINE Rects rects() const { return m_region.rects(); }
-    ALWAYS_INLINE bool isEmpty() const  { return !m_invalid && m_region.isEmpty(); }
+    Region region()
+    {
+        Region region;
+        for (const auto& rect : rects())
+            region.unite(rect);
+        return region;
+    }
+
+    ALWAYS_INLINE IntRect bounds() const { return m_minimumBoundingRectangle; }
+
+    // May return both empty and overlapping rects.
+    ALWAYS_INLINE const Rects& rects() const { return m_rects; }
+    ALWAYS_INLINE bool isEmpty() const  { return !m_invalid && m_rects.isEmpty(); }
     ALWAYS_INLINE bool isInvalid() const { return m_invalid; }
 
     void invalidate()
     {
         m_invalid = true;
-        m_region = Region();
+        m_rects = { };
     }
 
     ALWAYS_INLINE void add(const Region& region)
     {
         if (isInvalid())
             return;
-        m_region.unite(region);
-        mergeIfNeeded();
+        for (const auto& rect : region.rects())
+            add(rect);
```

Comment:
for each rectangle we invoke the (at first sight) non-trivial `add`. Can this really be `ALWAYS_INLINE`?

Reply:
yes, this whole `for` loop can be inlined, but the `add` itself cannot